asmcomp/s390x/emit.mlp

```diff
@@ -347,12 +366,34 @@ let emit_instr env i =
             `	brcl	15, {emit_symbol func}\n`
         end
 
-     | Lop(Iextcall { func; alloc; }) ->
-        if not alloc then emit_call func
-        else begin
+    | Lop(Iextcall {func; alloc; stack_ofs}) ->
+        if stack_ofs > 0 then begin
+          ` lgr {emit_reg reg_stack_arg_begin}, %r15\n`;
+          ` lgr {emit_reg reg_stack_arg_end}, %r15\n`;
+          ` agfi {emit_reg reg_stack_arg_end}, {emit_int stack_ofs}\n`;
+          emit_load_symbol_addr reg_r7 func;
+          emit_call "caml_c_call_stack_args";
+          `{record_frame env i.live (Dbg_other i.dbg)}\n`
+        end else if alloc then begin
           emit_load_symbol_addr reg_r7 func;
           emit_call "caml_c_call";
           `{record_frame env i.live (Dbg_other i.dbg)}\n`
+        end else begin
+          ` lgr %r1, %r15\n`;
+          cfi_remember_state ();
+          cfi_def_cfa_register "%r1";
+          (* NB: gdb has asserts on contiguous stacks that mean it
+             will not unwind through this unless we were to tag this
+             calling frame with cfi_signal_frame in it's definition.
+             Also, for some reason ocaml compiler uses %r12, can't use it here.
+             Just save stack pointer to temporary pointer and then to new stack *)
+          let offset = Domainstate.(idx_of_field Domain_c_stack) * 8 in
+          ` lg %r15, {emit_int offset}(%r10)\n`;
+          ` lay %r15, -168(%r15)\n`;
+          ` stg %r1, 160(%r15)\n`;
+          emit_call func;
+          ` lg %r15, 160(%r15)\n`;
+          cfi_restore_state ()
```

Comment:
No change necessary, but I wonder whether you could use r6 to save the OCaml stack pointer.  r6 is preserved by C functions and is marked as destroyed across calls in proc.ml (surprisingly?), so it might work.  Or use another C callee-save register that would have to be marked as destroyed across calls in proc.ml

Also: using @dra27's trick (#11846), the 160 bytes at the bottom of the C stack could possibly be reserved in advance, so that no stack adjustment would be needed here.  Again, this can be done later and should not block the merging of this PR.

Reply:
Is there any guarantee that %r6 wouldn't be silently used as well? %r12 is also marked as "preserved by C", but it didn't help. For that reason I went with safer option of using stack here.